addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go

```diff
@@ -0,0 +1,238 @@
+		log.Error(err, "unable to reconcile SecretGenControllerConfig")
+		return retResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SecretGenControllerConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&addonsv1alpha1.SecretGenControllerConfig{}).
+		Complete(r)
+}
+
+// ReconcileSecretGenControllerConfigNormal reconciles SecretGenControllerConfig data values secrets
+func (r *SecretGenControllerConfigReconciler) ReconcileSecretGenControllerConfig(
+	ctx context.Context,
+	log logr.Logger,
+	cluster *clusterapiv1beta1.Cluster,
+	clusterClient client.Client,
+	secretGenConfig *addonsv1alpha1.SecretGenControllerConfig) (_ ctrl.Result, retErr error) {
+
+	var (
+		patchCRD bool
+	)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(secretGenConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Patch SecretGenControllerConfig before returning the function
+	defer func() {
+		// patchCRD will be true if finalizer or ownerrefence is added
+		if patchCRD {
+			log.Info("Patching SecretGenControllerConfig")
+
+			if err := patchHelper.Patch(ctx, secretGenConfig.DeepCopy()); err != nil {
+				log.Error(err, "Error patching SecretGenControllerConfig")
+				retErr = err
+			}
+		}
+	}()
+
+	// If SecretGenControllerConfig is marked for deletion then delete the data value secret
+	if !secretGenConfig.GetDeletionTimestamp().IsZero() {
+		err := r.ReconcileSecretGenControllerConfigDelete(ctx, log, clusterClient, secretGenConfig, &patchCRD)
+		if err != nil {
+			log.Error(err, "Error reconciling SecretGenControllerConfig delete")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ReconcileSecretGenControllerConfigNormal(ctx, log, cluster, clusterClient, secretGenConfig, &patchCRD); err != nil {
+		log.Error(err, "Error reconciling SecretGenControllerConfig to create data value secret")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ReconcileSecretGenControllerConfigNormal reconciles SecretGenControllerConfig data values secret
+func (r *SecretGenControllerConfigReconciler) ReconcileSecretGenControllerConfigNormal(
+	ctx context.Context,
+	log logr.Logger,
+	cluster *clusterapiv1beta1.Cluster,
+	clusterClient client.Client,
+	secretGenConfig *addonsv1alpha1.SecretGenControllerConfig,
+	patchCRD *bool) (retErr error) {
+
+	// Add finalizer to addon secret
+	*patchCRD = util.AddFinalizerToCRD(log, constants.SecretGenControllerAddonName, secretGenConfig)
+
+	// add owner reference to addon secret
+	ownerReference := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               "Cluster",
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		Controller:         pointer.BoolPtr(true),
```

Comment:
question: any particular reason we want to add `controller:true` ? Are we going to enqueue cluster on `xxxConfig` deletion for garbage collection ?